### PR TITLE
Added deepseek_v3 in models

### PR DESCRIPTION
**Pull request by Sirorezka**
whoever for correct cp implementation 'softmax_scalling' factor must be added to ring attention.

---

> [!NOTE]
> **Medium Risk**
> Large new MoE/MLA path affects training numerics and distributed attention; ring CP may still need explicit `softmax_scale` forwarding per the PR note, so long-context CP parity is a follow-up risk.
> 
> **Overview**
> Adds a **PrimeRL custom `deepseek_v3` stack** so DeepSeek V3 can train with `impl = "custom"`: config, MLA + MoE modeling (grouped top‑k router, dense-then-MoE layers), flash/SDPA attention core, and **HF ↔ Prime** checkpoint key reshaping for routers, shared experts, and stacked expert weights.
> 
> **Wires the model into the trainer**: `AutoConfig` / `AutoModelForCausalLMPrimeRL` registration, ring and Ulysses context-parallel hooks patching `DeepSeekAttentionCore._compute_attention`, plus an example **`configs/deepseek_v3/sft.toml`** for a small test checkpoint on 2 GPUs.
> 
> **Tests** compare logits/grads and conversions against HuggingFace `DeepseekV3ForCausalLM`, with an optional (skipped) MATH KL sanity check on dual GPUs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7a02ed47519054bfd0961226b5fc46153551a8d6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        w1 = state_dict.pop(f"model.layers.{i}.mlp.experts.w1")
+        w2 = state_dict.pop(f"model.layers.{i}.mlp.experts.w2")
+        w3 = state_dict.pop(f"model.layers.{i}.mlp.experts.w3")
+
+        num_experts = w1.shape[0]
+
+        # bias is registred in HF Deepseek, but not used in the calculations
+        state_dict[f"model.layers.{i}.mlp.gate.e_score_correction_bias"] = torch.zeros(
+            num_experts
+        )
```
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py:186]: e_score_correction_bias is registred in HF Deepseek, but never used in the calculations or gradients. Thus it is always zero.

File: tests/unit/train/models/test_deepseek_v3.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+    original_method = DeepSeekAttentionCore._compute_attention
+
+    mock_group = MagicMock()
+    substitute_ring_attn(process_group=mock_group, heads_k_stride=1)
+
+    assert DeepSeekAttentionCore._compute_attention is not original_method
+
+    # Restore to avoid polluting other tests
+    DeepSeekAttentionCore._compute_attention = original_method
```
[the remark this answers is not in the document]
Comment (Sirorezka) [reply tests/unit/train/models/test_deepseek_v3.py:267]: fixed in new pr

File: src/prime_rl/trainer/models/deepseek_v3/modeling_deepseek_v3.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+
+    def init_buffers_post_meta(self):
+        pass
+
+
+__all__ = [
+    "DeepseekV3ForCausalLM",
+    "DeepseekV3Model",
+    "DeepseekV3PreTrainedModel",
+]
```
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/modeling_deepseek_v3.py:670]: What is KL validation table?

File: src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            attn_output = fa2_func(q, k, v, causal=True, softmax_scale=scale)
+        else:
+            # Varlen Attention
+            # inputs (bs==1, sl, nkv, d)
+            attn_output = self._compute_attention(
+                q[0], k[0], v[0], cu_seqlens, max_seqlen, softmax_scale=scale
+            )
+            attn_output = attn_output.unsqueeze(0)
+
+        return attn_output.contiguous()
```
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py:156]: Good point. will be fixed in next pr.

File: src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
+import torch
```
Comment (S1ro1) [inline src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py:1]: This file seems to copy most of the current attention impl without any (or few) changes, any reason for it? If there are any changes, let's move them to the shared impl if not breaking?

File: src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+        else:
+            # safetensors format might have experts splitted
+            for j in range(num_experts):
+                state_dict[f"model.layers.{i}.mlp.experts.{j}.gate_proj.weight"] = w1[j]
+                state_dict[f"model.layers.{i}.mlp.experts.{j}.down_proj.weight"] = w2[j]
+                state_dict[f"model.layers.{i}.mlp.experts.{j}.up_proj.weight"] = w3[j]
+
+
+def convert_hf_layer_to_prime(state_dict: dict[str, Tensor], layer_idx: int) -> None:
```
Comment (S1ro1) [inline src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py:204]: We require these methods to be implemented for RL, as those are used in weight broadcast. Need to implement this for the PR to be considered

File: configs/deepseek_v3/sft.toml [hunk 7]
```diff
@@ -0,0 +1,65 @@
+
```
Comment (S1ro1) [inline configs/deepseek_v3/sft.toml]: Can we change configs to full shape and also fix formatting?

**Review by S1ro1 — COMMENTED** [review]
Left some comments that are blocking currently, else looks reasonable to me. 

There are few blocking things - we now require 20 steps with kl_mismatch < 0.015 across all steps on math env with BS=64 before merging new models. Is this something you can do on your end? If not feel free to drop a config for it and I can run it at the earliest convenience. 

Also can you add/mention relevant parts to the model in docs/README where other model impls are mentioned.

File: src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py:186]: This however is required to be sent to inference server during RL.

File: src/prime_rl/trainer/models/layers/attn.py [hunk 8]
```diff
@@ -340,3 +340,7 @@ def _ring_compute_attention(self, q, k, v, cu_seqlens, max_seqlen):
     from prime_rl.trainer.models.qwen3_5_moe.modeling_qwen3_5_moe import Qwen3_5MoeGatedFlashAttention
 
     Qwen3_5MoeGatedFlashAttention._compute_attention = _ring_compute_attention
+
+    from prime_rl.trainer.models.deepseek_v3.attention_deepseek_v3 import DeepSeekAttentionCore
+
+    DeepSeekAttentionCore._compute_attention = _ring_compute_attention
```
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/models/layers/attn.py:346]: Seems relevant as well

File: src/prime_rl/trainer/models/deepseek_v3/modeling_deepseek_v3.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/models/deepseek_v3/modeling_deepseek_v3.py:670]: Explained in my comment

File: src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py:156]: Can we fix it in this so we don't pollute.

File: src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py (hunk 5 above)
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py:1]: I wasn't been able to reuse them, because 'FlashAttention' and 'SDPAAttention' have their own versions of q,k,v projections which are differ from the ones that are used in DeepSeek. Because of this to reuse original attention I would need to rewrite '__init__', 'forward' and 'attn_projections' methods. Which is basically the same as rewriting whole module from scratch.

File: configs/deepseek_v3/sft.toml (hunk 7 above)
Comment (Sirorezka) [reply configs/deepseek_v3/sft.toml]: Thx. Where can I find example of the full shape of the config?

File: src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/attention_deepseek_v3.py:156]: yeah, sorry, meant in next commit.

File: src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/converting_deepseek_v3.py:186]: fixed, added expert bias back

**Comment by Sirorezka** [comment]
Adding KL test results. Pls review when you'll have time @S1ro1 

Screenshot 2026-06-14 at 15 32 29

File: src/prime_rl/trainer/models/deepseek_v3/configuration_deepseek_v3.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                # convert to float
+                if key in rope_conf:
+                    rope_conf[key] = 1.0 * rope_conf[key]
+        else:
+            self.rope_type = "default"
+
+        self.rope_parameters = self.rope_scaling or self.rope_parameters
+        self.rope_parameters = (
+            self.rope_parameters if self.rope_parameters is not None else {}
+        )
```
[the remark this answers is not in the document]
Comment (Sirorezka) [reply src/prime_rl/trainer/models/deepseek_v3/configuration_deepseek_v3.py:127]: This is hallucination. 'test_empty_config' works without errors.

**Comment by Sirorezka** [comment]
image
Adding KL test results. Pls review when you'll have time @S1ro1
